### PR TITLE
Build: Replace globby with fdir

### PR DESCRIPTION
**Pull request by IanVS**
Issue:

## What I did

I was trying to upgrade our version of jest from 26 to 29, and I started getting failures of:

```
setImmediate is not defined
```

These were coming from `globby` (or one of its dependencies).  

I could have solved this in three ways, I think:

1) install a polyfill of setImmediate.  Seemed like the wrong approach.
2) configure those failing jest tests to run with a node environment rather than jsdom.  We may still want to explore this, but I need to figure out how to configure some of our tests to run with node environment and others with jsdom, since we have packages that expect to run in one or the other or both environments.
3) replace globby with something else that doesn't have this problem.

I did some checking around, and while `tiny-glob` seemed like a promising candidate, it didn't have a way to ignore node_modules (https://github.com/terkelg/tiny-glob/issues/79).

I then came across `fdir`, which claims to be the fastest glob solution around, with no dependencies and a small size.  https://github.com/thecodrr/fdir

This PR swaps out globby for fdir.  I think every thing we can do to gain some perf is a win, but happy to explore the other alternatives here as well.

## How to test

- [ ] Is this testable with Jest or Chromatic screenshots?
- [ ] Does this need a new example in the kitchen sink apps?
- [ ] Does this need an update to the documentation?

If your answer is yes to any of these, please make sure to include it in your PR.

### CODE REVIEW

**Comment by IanVS** [comment]
I ran fdir's benchmarks, which include fast-glob (globby) and tiny-glob, and it came out on top:

```
Running "Synchronous (1487 files, 230 folders)" suite...
Progress: 100%

  fdir 5.2.0 sync:
    90 ops/s, ±0.33%   | fastest

  glob sync:
    16 ops/s, ±0.62%   | slowest, 82.22% slower

  fast-glob sync:
    71 ops/s, ±0.67%   | 21.11% slower

  tiny-glob sync:
    45 ops/s, ±0.20%   | 50% slower

Finished 4 cases!
  Fastest: fdir 5.2.0 sync
  Slowest: glob sync

```

**Comment by IanVS** [comment]
Any idea what this pnp e2e failure could be from?

```
Error: Qualified path resolution failed: we looked for the following paths, but none could be accessed.
ERR! 
ERR! Source path: /tmp/storybook-e2e-testing/cra/.yarn/__virtual__/@storybook-addon-interactions-virtual-b7e2a61c70/4/root/.yarn/berry/cache/@storybook-addon-interactions-npm-7.0.0-alpha.34-d8e351c364-8.zip/node_modules/@storybook/addon-interactions/dist/cjs/preset/checkActionsLoaded
```

**Comment by IanVS** [comment]
I've broken out a separate PR to fix the issues with `addon-interactions`: https://github.com/storybookjs/storybook/pull/19334

File: code/lib/cli/scripts/generate-sb-packages-versions.js [hunk 1]
```diff
@@ -16,12 +21,7 @@ const run = async () => {
 
   if (!semver.valid(updatedVersion)) throw new Error(`Invalid version: ${updatedVersion}`);
 
-  const storybookPackagesPaths = await globby(
-    `${rootDirectory}/@(frameworks|addons|lib|renderers|presets)/**/package.json`,
-    {
-      ignore: '**/node_modules/**/*',
-    }
-  );
+  const storybookPackagesPaths = glob.sync();
```
Comment (ndelangen) [inline code/lib/cli/scripts/generate-sb-packages-versions.js:24]: Looks like this can be made async!

File: code/lib/codemod/src/index.js [hunk 2]
```diff
@@ -53,7 +53,12 @@ export async function runCodemod(codemod, { glob, logger, dryRun, rename, parser
     if (knownParser !== 'babel') inferredParser = extension;
   }
 
-  const files = await globby([glob, '!**/node_modules', '!**/dist']);
+  const files = new FDir()
+    .withFullPaths()
+    .glob(glob)
+    .exclude((p) => p.includes('/node_modules') || p.includes('/dist'))
+    .crawl(process.cwd())
+    .sync();
```
Comment (ndelangen) [inline code/lib/codemod/src/index.js:61]: this is actually running in an async function, so it should be possible to make it async?

File: code/lib/core-server/src/utils/StoryIndexGenerator.ts [hunk 3]
```diff
@@ -90,10 +90,12 @@ export class StoryIndexGenerator {
       this.specifiers.map(async (specifier) => {
         const pathToSubIndex = {} as SpecifierStoriesCache;
 
-        const fullGlob = slash(
-          path.join(this.options.workingDir, specifier.directory, specifier.files)
-        );
-        const files = await glob(fullGlob);
+        const files = new FDir()
```
Comment (ndelangen) [inline code/lib/core-server/src/utils/StoryIndexGenerator.ts:93]: also in an async function

File: code/lib/core-server/src/utils/watch-story-specifiers.ts [hunk 4]
```diff
@@ -66,16 +66,18 @@ export function watchStorySpecifiers(
...
+              // because the directory could already be within the files part (e.g. './x/foo/bar')
               path.basename(specifier.files)
             );
-            // glob only supports forward slashes
-            const files = await glob(dirGlob.replace(/\\/g, '/'));
+            const files = new FDir()
+              .withFullPaths()
+              .glob(globPattern)
+              .crawl(globDir)
+              .sync() as string[];
```
Comment (ndelangen) [inline code/lib/core-server/src/utils/watch-story-specifiers.ts:80]: async?

**Review by ndelangen — APPROVED** [review]
Though the code generally looks a bit more complex, I'm ok with it.

But i noticed in a few places we're using `.sync()`; which seems wasteful because this is going to touch the FS and so be potentially slow & blocking.

**Comment by IanVS** [comment]
Waiting for https://github.com/thecodrr/fdir/pull/80 to be released, which should fix yarn pnp. 

I'll try making these calls async as well.

**Review by ndelangen — APPROVED** [review]

**Comment by ndelangen** [comment]
Merge when green!

**Comment by IanVS** [comment]
Unfortunately I've hit another snag with fdir.  The way it handles symlinks is by resolved path, not the symlink path.  This breaks in our testing setup, where we symlink in various addon stories.  

The author is [considering](https://github.com/thecodrr/fdir/issues/23#issuecomment-1281810844) changing this behavior, but until/unless he does, I think we can't move forward with fdir.  I may need to put this back down for a while and focus on other parts of the jest upgrade, and come back to this later either when fdir's symlink behavior is changed, or find a different alternative, or maybe just find a way to adjust jest to work correctly with globby, our existing tool.

**Comment by IanVS** [comment]
I found a different way to handle the underlying issue with setImmediate.  We may want to re-evaluate the use of globby in the future, but given the challenges I've had getting fdir to work in an equivalent way, I'm closing this out for now.

**Comment by thecodrr** [comment]
@IanVS 

Would you be up for taking another look at this? There have been a lot of changes in the way `fdir` works. Especially everything around `globbing` is now much improved and works as expected.

A few examples:

1. Everything was rewritten in TypeScript
2. `fdir` now doesn't resolve symlinks by default (but requires `withSymlinks` option)
3. `globbing` now supports changing `picomatch` options with the `.globWithOptions` function. This fixes the behavior you mentioned in https://github.com/thecodrr/fdir/pull/85
4. A lot of other issues with globs were fixed.

The reason I am bringing this up again is due to the performance benefits Storybook can get by migrating to `fdir`. In your own benchmarks above, fdir quite a bit faster than the competition. If you like I can take a stab at renewing this PR.

**Comment by IanVS** [comment]
Thanks for your work on fdir, @thecodrr.  I think this PR is probably quite a bit outdated at this point, but I will try to find some time to take another crack at this in the not-too-distant future.

**Pull request closed without merging**